addons/stock/models/stock_quant.py

```diff
@@ -0,0 +1,497 @@
+            # it means that a negative quant has to be created as well.
+            negative_vals = vals.copy()
+            negative_vals['location_id'] = force_location_from and force_location_from.id or move.location_id.id
+            negative_vals['qty'] = float_round(-qty, precision_rounding=rounding)
+            negative_vals['cost'] = price_unit
+            negative_vals['negative_move_id'] = move.id
+            negative_vals['package_id'] = src_package_id
+            negative_quant_id = self.sudo().create(negative_vals)
+            vals.update({'propagated_from_id': negative_quant_id.id})
+
+        # In case of serial tracking, check if the product does not exist somewhere internally already
+        picking_type = move.picking_id and move.picking_id.picking_type_id or False
+        if lot_id and move.product_id.tracking == 'serial' and (not picking_type or (picking_type.use_create_lots or picking_type.use_existing_lots)):
+            if qty != 1.0:
+                raise UserError(_('You should only receive by the piece with the same serial number'))
+            other_quants = self.search([('product_id', '=', move.product_id.id), ('lot_id', '=', lot_id),
+                                        ('qty', '>', 0.0), ('location_id.usage', '=', 'internal')])
+            if other_quants:
+                lot_name = self.pool['stock.production.lot'].browse(lot_id).name
+                raise UserError(_('The serial number %s is already in stock') % lot_name)
+
+        # create the quant as superuser, because we want to restrict the creation of quant manually: we should always use this method to create quants
+        return self.sudo().create(vals)
+    # compatibility method
+    _quant_create = _quant_create_from_move
+
+    @api.multi
+    def _quant_split(self, qty):
+        self.ensure_one()
+        rounding = self.product_id.uom_id.rounding
+        if float_compare(abs(self.qty), abs(qty), precision_rounding=rounding) <= 0: # if quant <= qty in abs, take it entirely
+            return False
+        qty_round = float_round(qty, precision_rounding=rounding)
+        new_qty_round = float_round(self.qty - qty, precision_rounding=rounding)
+        # Fetch the history_ids manually as it will not do a join with the stock moves then (=> a lot faster)
+        self._cr.execute("""SELECT move_id FROM stock_quant_move_rel WHERE quant_id = %s""", (self.id,))
+        res = self._cr.fetchall()
+        new_quant = self.copy(default={'qty': new_qty_round, 'history_ids': [(4, x[0]) for x in res], 'qty': qty_round})
+        # self.write(cr, SUPERUSER_ID, quant.id, {'qty': qty_round}, context=context)
+        return new_quant
+
+    def _get_latest_move(self, cr, uid, quant, context=None):
+        move = False
+        for m in quant.history_ids:
+            if not move or m.date > move.date:
+                move = m
+        return move
```

Comment:
Could probably use max(quand.history_ids, key=date) (though it needs to handle the "no history" case)